lib/IRGen/IRGenDebugInfo.cpp

```diff
@@ -599,10 +605,24 @@ class IRGenDebugInfoImpl : public IRGenDebugInfo {
             : ~1ULL;
 
     // Handle Clang modules.
-    if (const clang::Module *ClangModule = Desc.getModuleOrNull()) {
+    if (ClangModule) {
       llvm::DIModule *Parent = nullptr;
-      if (ClangModule->Parent)
-        Parent = getOrCreateModule(*ClangModule->Parent);
+      if (ClangModule->Parent) {
+        // The loading of additional modules by Sema may trigger an out-of-date
+        // PCM rebuild in the Clang module dependencies of the additional
+        // module. A PCM rebuild causes the ModuleManager to unload previously
+        // loaded ASTFiles. For this reason we must use the cached ASTFile
+        // information here instead of the potentially dangling pointer to the
+        // ASTFile that is stored in the clang::Module object.
```

Comment:
This doesn't make sense to me. Wouldn't this be just as dangling as what Clang has? Wouldn't it be *more* dangling? Wouldn't other parts of the Swift AST be invalidated if Clang unloads a loaded ASTFile?

Reply:
I'm pretty sure that `IRGenDebugInfo` is the only place that asks `clang::Module`s questions about the ASTFiles. `ModuleManager::removeModules()` deletes ASTFiles, but as far as I can tell, it doesn't know anything about the `clang::Module` that may have pointers to them.

https://github.com/llvm/llvm-project/blob/1bccafe4f6f2b8e6496641e55cccb2dab2da58e2/clang/lib/Serialization/ModuleManager.cpp#L219